src/core/mapparser.cpp

```diff
@@ -0,0 +1,116 @@
+#include "mapparser.h"
+#include "config.h"
+#include "log.h"
+#include "project.h"
+
+MapParser::MapParser()
+{
+
+}
+
+MapLayout *MapParser::parse(QString filepath, bool *error, Project *project)
+{
+    QFile file(filepath);
+    if (!file.open(QIODevice::ReadOnly)) {
+        *error = true;
+        logError(QString("Could not open Advance Map 1.92 Map .map file '%1': ").arg(filepath) + file.errorString());
+        return nullptr;
+    }
+
+    QByteArray in = file.readAll();
+    file.close();
+
+    if (in.length() < 20 || in.length() % 2 != 0) {
+        *error = true;
+        logError(QString("Advance Map 1.92 Map .map file '%1' is an unexpected size.").arg(filepath));
+        return nullptr;
+    }
+
+    int borderWidth = static_cast<unsigned char>(in.at(16)); // 0 in RSE .map files
+    int borderHeight = static_cast<unsigned char>(in.at(17)); // 0 in RSE .map files
+    int numBorderTiles = borderWidth * borderHeight; // 0 if RSE
+
+    int mapDataOffset = 20 + (numBorderTiles * 2); // FRLG .map files store border metatile data after the header
+    int mapWidth = static_cast<unsigned char>(in.at(0)) |
+                   (static_cast<unsigned char>(in.at(1)) << 8) |
+                   (static_cast<unsigned char>(in.at(2)) << 16) |
+                   (static_cast<unsigned char>(in.at(3)) << 24);
+    int mapHeight = static_cast<unsigned char>(in.at(4)) |
+                    (static_cast<unsigned char>(in.at(5)) << 8) |
+                    (static_cast<unsigned char>(in.at(6)) << 16) |
+                    (static_cast<unsigned char>(in.at(7)) << 24);
+    int mapPrimaryTilesetNum = static_cast<unsigned char>(in.at(8)) |
+                               (static_cast<unsigned char>(in.at(9)) << 8) |
+                               (static_cast<unsigned char>(in.at(10)) << 16) |
+                               (static_cast<unsigned char>(in.at(11)) << 24);
+    int mapSecondaryTilesetNum = static_cast<unsigned char>(in.at(12)) |
+                                 (static_cast<unsigned char>(in.at(13)) << 8) |
+                                 (static_cast<unsigned char>(in.at(14)) << 16) |
+                                 (static_cast<unsigned char>(in.at(15)) << 24);
+
+    /*int maxMetatiles = primaryTileset ? Project::getNumMetatilesPrimary() : Project::getNumMetatilesTotal() - Project::getNumMetatilesPrimary();
```

Comment:
Remove dead code?

Reply:
Yeah, I want to clean this part up by doing maybe a few more proper checks to see if the input is valid. I'll look into it.